atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStore.java

```diff
@@ -0,0 +1,151 @@
+
+    public CassandraTimestampStore(CassandraKeyValueService cassandraKeyValueService) {
+        this.cassandraKeyValueService = cassandraKeyValueService;
+    }
+
+    /**
+     * Creates the timestamp table. Note that this operation is idempotent.
+     */
+    public void createTimestampTable() {
+        cassandraKeyValueService.createTable(
+                AtlasDbConstants.TIMESTAMP_TABLE,
+                CassandraTimestampUtils.TIMESTAMP_TABLE_METADATA.persistToBytes());
+    }
+
+    /**
+     * Gets the upper timestamp limit from the database if it exists.
+     * @return Timestamp limit stored in the KVS, if present
+     * @throws IllegalStateException if the timestamp limit has been invalidated
+     */
+    public synchronized Optional<Long> getUpperLimit() {
+        return cassandraKeyValueService.clientPool.runWithRetry(client -> {
+            ByteBuffer queryBuffer = CassandraTimestampUtils.constructSelectTimestampBoundQuery();
+            try {
+                CqlResult result = client.execute_cql3_query(queryBuffer, Compression.NONE, ConsistencyLevel.QUORUM);
+                List<CqlRow> cqlRows = result.getRows();
+                if (cqlRows.isEmpty()) {
+                    return Optional.<Long>empty();
+                }
+                return Optional.of(CassandraTimestampUtils.getLongValue(result));
+            } catch (TException e) {
+                throw Throwables.rewrapAndThrowUncheckedException(e);
+            }
+        });
+    }
+
+    /**
+     * Stores an upper timestamp limit in the database in an atomic way.
+     * @param expected Expected current value of the timestamp bound (or null if there is no bound)
+     * @param target Upper timestamp limit we want to store in the database
+     * @throws ConcurrentModificationException if the expected value does not match the bound in the database
+     */
+    public synchronized void storeTimestampBound(Long expected, long target) {
+        cassandraKeyValueService.clientPool.runWithRetry(client -> {
+            ByteBuffer queryBuffer = CassandraTimestampUtils.constructCheckAndSetCqlQuery(expected, target);
+            try {
+                CqlResult result = client.execute_cql3_query(queryBuffer, Compression.NONE, ConsistencyLevel.QUORUM);
+                checkOperationWasApplied(expected, target, result);
+                return null;
+            } catch (TException e) {
+                throw Throwables.rewrapAndThrowUncheckedException(e);
+            }
+        });
+    }
+
+    /**
+     * Writes a backup of the existing timestamp to the database. After this backup, this timestamp service can
+     * no longer be used until a restore.
+     *
+     * This may be implemented as follows:
+     *  - Read the value of the backup timestamp. If this is readable, skip (already backed up).
+     *  - Read the value of the timestamp (TS). If this is unreadable, fail.
+     *  - Do a conditional logged batch:
```

Comment:
Should bear in mind that atomicity can't be guaranteed, so you might have to deal with CAS success and PUE failure.